### ISSUE TITLE
State migration

### ISSUE THREAD

**Issue by bkchr** [polkadot-fellows/runtimes#74]
We need to migrate all parachains and relay chain to state version 1. There is a pallet for doing this. With 1.0.0 Kusama will enable the state migration. After that we also need to migrate the parachains and Polkadot and its parachains. This issue works as a tracking issue.

**Comment by bkchr** [comment]
@cheme could you please post some status on what needs to be done for Kusama? Aka does the migration works on its own or is that driven by some offchain thingy? 

After Kusama was successful, we should also directly start doing it for Polkadot.

**Comment by cheme** [comment]
Looks ready to go with next kusuma runtime.

This line set at 1 (switch to hybrid state when new runtime released) : https://github.com/polkadot-fellows/runtimes/blob/94b2798b69ba6779764e20a50f056e48db78ebef/relay/kusama/src/lib.rs#L146

Start of migration added to Unreleased https://github.com/polkadot-fellows/runtimes/blob/94b2798b69ba6779764e20a50f056e48db78ebef/relay/kusama/src/lib.rs#L1731

Unrealeased set as runtime migration https://github.com/polkadot-fellows/runtimes/blob/94b2798b69ba6779764e20a50f056e48db78ebef/relay/kusama/src/lib.rs#L1654C35-L1654C35

Not that if there is many migration runing together, might be an idea to lower the limit per block:
https://github.com/polkadot-fellows/runtimes/blob/94b2798b69ba6779764e20a50f056e48db78ebef/relay/kusama/src/lib.rs#L2714

here up to 4800 item or 408000 octet which in case of max out blocks can add :
db_weight_reads_writes(1, 1) = (25_000   + 100_000) * 1000.
(using rocksdb cst).
`125_000_000 * 4_800 + 408_000 *  1_139 + some_fix_weight` 
`600 * 10^9 + 464_712_000` out of 2_000_000_000_000, so if I check correctly about <U>**1 third**</U> of a block weight.
Note that in case of relay chain, consuming weight is not a must have from my point of view.


https://github.com/polkadot-fellows/runtimes/blob/94b2798b69ba6779764e20a50f056e48db78ebef/relay/kusama/src/lib.rs#L1468


Warning if the start line get remove from unreleased, the line 146 must be set to 0 (to avoid hybrid state).

**Comment by cheme** [comment]
@bkchr the weight use in each block is  an important point that should be in release note I think.

As a relay chain we could also ignore this weight to prevent any issue.

**Comment by cheme** [comment]
actually I did use 2 * 10 ^ 12 for block weight but on relay it may be 6*10^12 so would not be worrying then

**Comment by bkchr** [comment]
The migration will run entirely on chain and doesn't require any external interactions?

**Comment by cheme** [comment]
> chain and doesn't require any exte

on chain, no possible external interactions.

**Comment by ggwpez** [comment]
Kusama is done? I queried `StateTrieMigration.MigrationProcess` and got:  
```pre
 {
  "progress_top": {
    "name": "Complete",
    "values": []
  },
  "progress_child": {
    "name": "ToStart",
    "values": []
  },
  "size": 239581990,
  "top_items": 881284,
  "child_items": 1373
}
```
The [Events](https://kusama.subscan.io/event?address=&module=statetriemigration&event=all&startDate=&endDate=&startBlock=&endBlock=&timeType=date&version=1000000) also look fine. Just ~480 blocks to migrate everything? Looks like it did about 4800 keys in some blocks 😳

**Comment by cheme** [comment]
4800 was the limit indeed.
This rpc can be call (https://github.com/paritytech/cumulus/pull/1424) to double check the state did migrate correctly (requires runing it locally as it is unsafe).
Ultimately another good check is to run a warp sync (if the state is not migrated warp sync should not be working).
(cannot run these right now I am ooo)

**Comment by ggwpez** [comment]
>Ultimately another good check is to run a warp sync (if the state is not migrated warp sync should not be working).

I just tried and it still works.

**Comment by bkchr** [comment]
@cheme so did it finished successfully?

**Comment by cheme** [comment]
Yes, but if warp sync did pass, all is fine.
(warp sync is a guaranteed fail during migration, and the counter on chain do state the migration is finished).

**Comment by bkchr** [comment]
@cheme can you prepare the changes for Polkadot?

**Comment by cheme** [comment]
Will do (tomorrow most likely), I think since block time is longer on polkadot, keeping the same config as kusuma should be fine.

**Comment by bkchr** [comment]
Ping @cheme

**Comment by bkchr** [comment]
BTW, we also need to migrate the system chains.

**Comment by cheme** [comment]
https://github.com/polkadot-fellows/runtimes/pull/170
systems chain will be another beast.
I was never really happy with the rpc calls process manual migration, last time I thought about it I was thinking of just running the automatic process, maybe with an exclusion list (first scan offchain for big value and process them one by one, but then there is still the issue of big value being added to the chain beetween the scan and the start of the migration, but realistically speaking and with a bit of knowledge of the system chain logic we can probably be confident this would not happen (big problematic value should be rather rare and no sane chain would create them randomly).

But the rpc approach can probably do the job, just the amount of energy to manage them worries me.
(with automatic, we need to contact someone competant to know where can be big problematic value, do a scan of the state to find the existing one and add a skip list of them (it is actually a process first then skip list).
With is automatic approach I would need to add some new code to the migration pallet (the skip list related code) though.
cc @kianenigma

**Comment by cheme** [comment]
> I thought this was also just some bot doing this? Or what you mean by manual in this case?

yes manual by a bot, still need to run the bot (need slashable fee deposit, also I am not sure anymore if it should target a specific account (seems like a liability: should be open to everyone, but I don't remember how we ensure a single call is done per blocks).
Maybe it is fine.

**Comment by bkchr** [comment]
I mean just opening this for one account sounds fine to me. I mean we are speaking here about a one time migration. 

> but I don't remember how we ensure a single call is done per blocks

This could be done with a storage value that is set to `true` when the call was done and removed at `on_finalize`.

**Comment by kianenigma** [comment]
> Kusama is done? I queried `StateTrieMigration.MigrationProcess` and got:
> 
> ```
>  {
>   "progress_top": {
>     "name": "Complete",
>     "values": []
>   },
>   "progress_child": {
>     "name": "ToStart",
>     "values": []
>   },
>   "size": 239581990,
>   "top_items": 881284,
>   "child_items": 1373
> }
> ```
> 
> The [Events](https://kusama.subscan.io/event?address=&module=statetriemigration&event=all&startDate=&endDate=&startBlock=&endBlock=&timeType=date&version=1000000) also look fine. Just ~480 blocks to migrate everything? Looks like it did about 4800 keys in some blocks, nice 😳

So the kusama state is 240 MiB? interesting. I am still missing a tool like https://github.com/paritytech/polkadot-sdk/issues/449, I wonder if there is an ecosystem tool for this that I am not aware of?

**Comment by kianenigma** [comment]
@cheme I would not use the automatic migration on a system chain, as any error will likely cause the parachain to stop. 

Maybe on a kusama system chain, but 100% not for Polkadot. I wrote a TS bot that still should work fine to trigger the migrations one by one, and it should all be free. Have we ever used the signed migration?

**Comment by cheme** [comment]
> @cheme I would not use the automatic migration on a system chain, as any error will likely cause the parachain to stop.
> 
> Maybe on a kusama system chain, but 100% not for Polkadot. I wrote a TS bot that still should work fine to trigger the migrations one by one, and it should all be free. Have we ever used the signed migration? ref: https://github.com/paritytech/polkadot-scripts/blob/master/src/services/state_trie_migration.ts

Long time ago we did some with @PierreBesson when doing rococo and westand (I think statemine). But for me it is a bit too long ago.

**Comment by bkchr** [comment]
After Polkadot is done, we need to work on the parachains. As said above, I don't see any real problem in using the offchain bot.

**Comment by ggwpez** [comment]
I added a list to the issue description. Please tick of the ones that are done.

**Comment by bkchr** [comment]
Kusama is already done. Polkadot should be finished after the next runtime upgrade.

**Comment by cheme** [comment]
> @cheme can you please also add it for the missing ones?

is there new ones? (I mean asset-hub and collectives were done).

**Comment by ggwpez** [comment]
![Screenshot 2024-05-13 at 12 59 27](https://github.com/polkadot-fellows/runtimes/assets/10380170/187ed2bc-92c7-403c-8bd5-1d7f20a99f9c)

Yea there are some remaining it looks like. Asset-Hub Kusama has it behind a feature-gate that was never enabled i guess?

**Comment by kianenigma** [comment]
PAH and KAH done in https://github.com/polkadot-fellows/runtimes/pull/604 by @nkpar.

**Comment by bkchr** [comment]
So only collectives and encointer are missing.

**Comment by bkchr** [comment]
@kianenigma @nkpar are you sure that you finished the migration?
```
{
  progressTop: {
    LastKey: 0x682a59d51ab9e48a8c8cc418ff9708d2b99d880ec681799c0cf30e8886371da928210239547ef1b1a87a0b1c87bd80bc1e000000f24c6002d53033fdb074b26a1c46b645f2de5ff8d00ab2a53b65e70603a09de0bec205589fd32cedda051172cf4947f3
  }
  progressChild: ToStart
  size_: 225,689,399
  topItems: 4,131,097
  childItems: 0
}
```

While there are probably no child tries to migrate, it doesn't look like the migration was finished?

**Comment by bkchr** [comment]
Why is the block that slow on sending these transactions? Sometimes there is only one tx per minute.
